listenstore/listenstore/listenstore.py

```diff
@@ -188,24 +193,70 @@ def get_playing_now(self, user_id):
 
 REDIS_INFLUX_USER_LISTEN_COUNT = "ls.listencount." # append username
 class InfluxListenStore(ListenStore):
+
+    # variables which define how much time to cache listen count in redis
+    LISTENCOUNT_CACHE_TIME_LOW = 15 * 60 # in seconds. 15 minutes
+    LISTENCOUNT_CACHE_TIME_MEDIUM = 60 * 60 # in seconds. 1 hour
+    LISTENCOUNT_CACHE_TIME_HIGH = 24 * 60 * 60 # in seconds. 1 day
+    LOW_LISTEN_THRESHOLD = 1000
+    MEDIUM_LISTEN_THRESHOLD = 10000
+
     def __init__(self, conf):
         ListenStore.__init__(self, conf)
         self.redis = Redis(conf['REDIS_HOST'])
         self.influx = InfluxDBClient(host=conf['INFLUX_HOST'], port=conf['INFLUX_PORT'], database=conf['INFLUX_DB_NAME'])
 
 
-    def _get_num_listens_for_user(self, user_name):
-        count = self.redis.get(REDIS_INFLUX_USER_LISTEN_COUNT + user_hash(user_name))
-        if count:
-            return int(count)
+    def get_listen_count_for_user(self, user_name, need_exact = False):
+        """ Returns total listen count for user with user_name passed.
 
-        results = self.influx.query("""SELECT count(*) 
-                                         FROM listen 
-                                        WHERE user_name = '%s'""" % (user_name, from_ts, to_ts, limit))
-        result = results.get_points(measurement='listen')
-#        count = self.redis.get(REDIS_INFLUX_USER_LISTEN_COUNT + user_hash(user_name))
+            Args:
+                user_name: user name of user whose listen count is to be found
+                need_exact: signifies if cached value in redis will suffice or if exact value is needed
+        """
 
-        return 0
+        if not need_exact:
+            # check if the user's listen count is already in redis
+            # if already present return it directly instead of calculating it again
+            count = self.redis.get(REDIS_INFLUX_USER_LISTEN_COUNT + user_name)
+            if count:
+                return int(count)
+
+        try:
+            results = self.influx.query("""SELECT count(*)
+                                             FROM listen
+                                            WHERE user_name = '%s'""" % (user_name))
+        except (InfluxDBServerError, InfluxDBClientError) as e:
+            self.log.error("Cannot query influx: %s" % str(e))
+            raise
+
+        # get the number of listens from the json
+        try:
+            count = results.raw['series'][0]['values'][0][1]
+        except KeyError:
+            count = 0
+
+        # put this value into redis with expiry time based on the number of listens
+        # that the user has
+        if count <= InfluxListenStore.LOW_LISTEN_THRESHOLD:
+            self.redis.setex(REDIS_INFLUX_USER_LISTEN_COUNT + user_name, count, InfluxListenStore.LISTENCOUNT_CACHE_TIME_LOW)
```

Comment:
Is this key a string? I'd much prefer that it uses string interpolation instead of concatenation.

This key is used 3 times in 6 lines. You should put it in a variable.